src/coreclr/vm/jithelpers.cpp

```diff
@@ -572,6 +572,38 @@ FORCEINLINE INT64 FastDbl2Lng(double val)
 #endif
 }
 
+//------------------------------------------------------------------------
+// TruncateDouble: helper function to truncate double 
+//                 numbers to nearest integer (round towards zero).
+//
+// Arguments:
+//    val  - double number to be truncated.
+//
+// Return Value:
+//    double: truncated number (rounded towards zero)
+// 
+double TruncateDouble(double val)
```

Comment:
Is there an user-observable difference with this new version? I.e., is this a breaking change?

@jkotas Is this ok from a VM coding conventions standpoint? i.e., a "loose" function, marked with `FCALL_CONTRACT`?

Reply:
There will be a user observable difference  for ```double -> ulong``` cast. The older version was not IEEE 754 compliant and since the new AVX512 instructions are IEEE 754 compliant, we had to change the existing implementation for consistency between AVX512 and non AVX512 machines.

Reply:
Its a user observable difference on x86/x64. It already differed from other platforms and is now being made correct.

The behavior on overflow is likewise becoming consistent with x86/x64 in general, although still differs from other platforms (xarch returns a sentinel, arm saturates).

We have an approved, but NYI, issue to make all platforms consistent for these overflow scenarios (which are considered UB).

Reply:
Can you remind me what is the uniform deterministic overflow behavior we agreed on? Should this change work towards that?

Reply:
We agreed to do saturation to match Arm64 and what WASM requires.

Moving towards this is generally desirable but is quite a bit more work overall. The impact will be lesser on AVX-512 capable hardware due to vfixupimm, but still needs some work for down level to ensure it doesn’t have a significant impact for sse4.1+ hardware

I think it would be better to do that all at once, likely be reviving the PR I had open around 6 months back. It had been blocked at the time due to some mono changes that were required which I believe have now been completed

Reply:
That is, I think it would be good to take this or which fixes a clear bug on x64 and improves the perf on more capable hardware

man’s then to independently take the changes required to make all conversions deterministic

Reply:
> still needs some work for down level to ensure it doesn’t have a significant impact for sse4.1+ hardware

Looking at the code, this PR may require work to mitigate the perf impact for non-AVX512 too. I will wait for answer to https://github.com/dotnet/runtime/pull/84384#discussion_r1239298409.

If we are going to perf tune the implementation on different types of hardware, my thinking is that we can switch to the final desirable behavior for this conversion first so that we go through the perf tuning just once.

Reply:
@tannergooding @jkotas Here. we are aiming towards making the older JITDbl2Ulng IEEE 754 compliant for x86/x64 machines. To use the newer available Conversion instruction (i.e. vcvtsd2usi), which is IEEE 754 compliant, we had to change the old behavior to maintain consistency. The new instructions saturate if overflow occurs. 

Is this behavior inline with the uniform deterministic behavior we are talking about?

Reply:
@tannergooding Was your earlier comment about this PR returning a sentinel an oversight?

Reply:
> Was your earlier comment about this PR returning a sentinel an oversight?

No, the AVX512F instruction doesn't quite saturate. Rather it is documented as:
> If a converted result cannot be represented in the destination format, the floating-point invalid exception is raised, and if this exception is masked, the integer value `2^w – 1` is returned, where w represents the number of bits in the destination format.

That means `Avx512F.ConvertToUInt32WithTruncation(Vector128.CreateScalar(4294967296.0f))` and `Avx512F.ConvertToUInt32WithTruncation(Vector128.CreateScalar(-2.0f))` both return `4294967295`, where-as true saturation would have the latter return `0` instead.

This is a different sentinel than what is returned for signed conversions, which is `-2147483648` (int) or `-9223372036854775808` (long).

That is, xarch has signed conversions return `MinValue` as the sentinel (for out of range conversions) while unsigned conversions return `MaxValue` as the sentinel.

------------

Also notably IEEE 754 doesn't define the behavior for such out of bounds conversions. It is strictly implementation defined behavior, outside of signaling the `invalid operation exception` if IEEE 754 exception handling is enabled:
> When a NaN or infinite operand cannot be represented in the destination format and this cannot otherwise
be indicated, the invalid operation exception shall be signaled. When a numeric operand would convert to
an integer outside the range of the destination format, the invalid operation exception shall be signaled if
this situation cannot otherwise be indicated.

We agreed on normalizing to saturation because:
1. This is what Arm64 does and what WASM requires. It's what many other languages have standardized on as well
2. Xarch hardware has instructions that allow it to more easily handle these edge cases, this is particularly true for AVX512 and later (1 additional instruction). But it is still fairly cheap on older hardware, particularly for unsigned where its just `maxss 0, value; convert`. For signed it's a bit harder but there are still several options available that only take a couple instructions.

Reply:
> Looking at the code, this PR may require work to mitigate the perf impact for non-AVX512 too.

If we were to improve this, we should handle it intrinsically and avoid the helper altogether.

For `AVX512` saturation of unsigned is just the following, which is around 4 extra cycles (so 11 total, rather than the 7 it would be without saturation handling):
```
vxorps tmp, tmp, tmp     ; zero tmp
vmaxss tmp, value, tmp   ; tmp = max(value, tmp), taking tmp if value is NaN
vcvttss2usi dst, tmp     ; IEEE 754 compliant unsigned conversion, returning uint.MaxValue if out of range
```

For pre-avx512, its similar and is around 7 extra cycles (this is 11 cycles more than AVX-512):
```
vxorps tmp, tmp, tmp     ; zero tmp
vmaxss tmp, value, tmp   ; tmp = max(value, tmp), taking tmp if value is NaN
vcvtss2sd tmp, tmp       ; Upcast from float to double
vcvtsd2si dst, tmp       ; IEEE 754 compliant signed conversion, returning long.MinValue if out of range
shr dst, 63              ; Propagate sign bit down to all bits, converting long.MinValue to `uint.MaxValue`
```

-------------------

Signed conversions are slightly more complex as we also have to handle two edge cases, but they are similarly "cheap" and can be handled using `vfixupimm`, `vrange`, or a combination of `vmax`/`vmin`, depending on whether we're targeting AVX-512, SSE4.1, or baseline hardware.

Reply:
And, of course, the approved changes (https://github.com/dotnet/runtime/issues/61885) cover APIs giving users access to platform native conversions where the behavior may remain undefined, for if they do need the perf and are in a scenario where the edge cases don't matter or were handled and the JIT couldn't handle it themselves.

Reply:
@tannergooding So as far as I understood, we are good to go with the changes. Right? There is some future work required to make sure we standardize to saturation but at this point, these changes will help us in that direction.

The AVX512F instructions will not saturate (Follow the expected saturation behavior and represents it as a floating point exception with max value), but this works in the favor of what we want achieve as the end goal https://github.com/dotnet/runtime/pull/84384#discussion_r1240202865 and https://github.com/dotnet/runtime/pull/84384#discussion_r1240565719